sdk/program/src/loader_upgradeable_instruction.rs

```diff
@@ -0,0 +1,75 @@
+//! Instruction types
+
+use crate::pubkey::Pubkey;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum UpgradeableLoaderInstruction {
+    /// Write program data into a source account.
+    ///
+    /// A source account is an intermediary that once fully populated is used
+    /// with the `Finalize` instruction to populate the program's code account.
+    ///
+    /// # Account references
+    ///   0. [writable, signer] source account to write program data to.
+    Write {
+        /// Offset at which to write the given bytes
+        offset: u32,
+
+        /// Serialized program data
+        #[serde(with = "serde_bytes")]
+        bytes: Vec<u8>,
+    },
+
+    /// Deploy an executable program.
+    ///
+    /// A program consists of an executable account and code account pair.
+    ///   - The executable account's address will serve as the program id in
+    ///     instructions that execute this program.
+    ///   - The code account will remain mutable by the loader only and holds
+    ///     the program data and authority information.
+    ///
+    /// The `Deploy` instruction requires no signers and MUST be included within
+    /// the same Transaction as the system program's `CreateAccount` instruction
+    /// that creates the executable and code accounts. Otherwise another party
+    /// can acquire ownership of the uninitialized accounts.
+    ///
+    /// # Account references
+    ///   0. [writable] The code account.
+    ///   1. [writable] The executable account.
+    ///   2. [writable] The source account where the program data has been
+    ///      written.
+    ///   3. [] The program's authority.
```

Comment:
```suggestion
    ///   3. [] The program's upgrade authority.
```
For flexibility we should support `None` here.  I could imagine re-implementing `solana deploy` to use the upgradable loader in all cases if we can make it a superset of the existing loader